aioxmpp/utils.py

```diff
@@ -344,3 +348,217 @@ def to_nmtoken(rand_token):
+        self.on_deadtime_hard_limit_tripped()
+
+    def _retrigger_timers(self):
+        now = time.monotonic()
+
+        if self._soft_limit_timer is not None:
+            self._soft_limit_timer.cancel()
+            self._soft_limit_timer = None
+
+        if self._soft_limit is not None:
+            self._soft_limit_timer = self._loop.call_later(
+                self._soft_limit.total_seconds() - (now - self._last_rx),
+                self._trip_soft_limit
+            )
+
+        if self._hard_limit_timer is not None:
+            self._hard_limit_timer.cancel()
+            self._hard_limit_timer = None
+
+        if self._hard_limit is not None:
+            self._hard_limit_timer = self._loop.call_later(
+                self._hard_limit.total_seconds() - (now - self._last_rx),
+                self._trip_hard_limit
+            )
+
+    def _reset_trips(self):
+        self._soft_limit_tripped = False
+        self._hard_limit_tripped = False
+        self._last_rx = time.monotonic()
+
+    def notify_received(self):
+        """
+        Inform the aliveness check that something was received.
+
+        Resets the internal soft/hard limit timers.
+        """
+        self._reset_trips()
+        self._retrigger_timers()
+
+    @property
+    def deadtime_soft_limit(self):
+        """
+        Soft limit for the timespan in which no data is received in the stream.
+
+        When the last data reception was longer than this limit ago,
+        :meth:`on_deadtime_soft_limit_tripped` emits once.
+
+        Changing this makes the monitor re-check its limits immidately. Setting
+        this to :data:`None` disables the soft limit check.
+
+        Note that setting this to a value greater than
+        :attr:`deadtime_hard_limit` means that the hard limit will fire first.
+        """
+        return self._soft_limit
+
+    @deadtime_soft_limit.setter
+    def deadtime_soft_limit(self, value):
+        if self._soft_limit_timer is not None:
+            self._soft_limit_timer.cancel()
+        self._soft_limit = value
+        self._retrigger_timers()
+
+    @property
+    def deadtime_hard_limit(self):
+        """
+        Hard limit for the timespan in which no data is received in the stream.
+
+        When the last data reception was longer than this limit ago,
+        :meth:`on_deadtime_hard_limit_tripped` emits once.
+
+        Changing this makes the monitor re-check its limits immidately. Setting
+        this to :data:`None` disables the hard limit check.
+
+        Note that setting this to a value less than
+        :attr:`deadtime_soft_limit` means that the hard limit will fire first.
+        """
+        return self._hard_limit
+
+    @deadtime_hard_limit.setter
+    def deadtime_hard_limit(self, value):
+        if self._hard_limit_timer is not None:
+            self._hard_limit_timer.cancel()
+        self._hard_limit = value
+        self._retrigger_timers()
+
+
+class proxy_property:
```

Comment:
Why not implement this as a function which provides property with the appropriate functions? Along the lines:
```python
def proxy_property(owner_attr, member_attr, *, readonly=False, allow_delete=False, doc=None):
      getter = lambda self: getattr(getattr(self, owner_attr), member_attr)
      setter = None if readonly else lambda self, value: ...
      ....
      return property(getter, setter, deleter, doc)
```
IMO this is the more readable variant.

One could even think about compiling the lambdas from source to avoid the lookup of getattr for better performance (of course at the cost of worse error detection if the attr strings contain non-indentifiers, so probably not a good idea).

Reply:
I don’t think the lambdas are more readable, and building source code is even less of an option. I went the "building source code" route with xsltea, and I’d like to never go down that rabbithole again.

Reply:
well then don't use lambdas but local functions, then one can even use the decorator form of property – I just somehow prefer reusing property instead of rolling a weird property-like descriptor.

Reply:
Using local functions makes it structurally very similar to a class then.

I don’t see value in forcing a re-use of property in this case. This variant can easily be extended to fetch a docstring if a type for the member is given, too (helping with sphinx). I’m not sure how easy that’s possible with a property wrapper.

Reply:
I found a reason to re-use ``property``: It provides the same interface to override setters or getters in derived classes—thus making it more transparent in its use.

Reply:
Done.